providers/sadad/sadad.go

```diff
@@ -0,0 +1,138 @@
+package sadad
+
+import (
+	"context"
+	"crypto/aes"
+	b64 "encoding/base64"
+	"github.com/GoFarsi/paygap/client"
+	"github.com/GoFarsi/paygap/status"
+	"github.com/andreburgaud/crypt2go/ecb"
+	"google.golang.org/grpc/codes"
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	baseUrl           = "https://sadad.shaparak.ir"
+	requestPaymentUrl = "/api/v0/Request/PaymentRequest"
+	requestVerifyUrl  = "/api/v0/Advice/Verify"
+)
+
+func New(client client.Transporter, terminalId string, amount int64,
+	merchantKey string, returnUrl string, merchantId string,
+	purchasePage string, enableMultiplexing bool,
+	multiplexingData *MultiplexingData) (*Sadad, error) {
+
+	if client == nil {
+		return nil, status.ERR_CLIENT_IS_NIL
+	}
+	orderId := string(rand.Int())
+	str := []string{terminalId, orderId, string(amount)}
+	joinedString := strings.Join(str, ";")
+
+	sadad := &Sadad{
+		Client:             client,
+		TerminalId:         terminalId,
+		MerchantId:         merchantId,
+		Amount:             amount,
+		OrderId:            orderId,
+		AdditionalData:     "",
+		LocalDateTime:      time.Now(),
+		ReturnUrl:          returnUrl,
+		SignData:           "",
+		EnableMultiplexing: enableMultiplexing,
+		MultiplexingData:   *multiplexingData,
+		MerchantKey:        merchantKey,
+		PurchasePage:       purchasePage,
+	}
+
+	signedDataAsSadadWay, er := sadad.SigningData(joinedString)
+	if er != nil {
+		panic(er)
+	}
+	sadad.SignData = signedDataAsSadadWay
+
+	return sadad, nil
+}
+
+func (s *Sadad) PayRequest(ctx context.Context) (*PayResultData, error) {
```

Comment:
In PayRequest better get `PaymentRequest` as parameter.

Reply:
Thanks. Done.